### PR TITLE
Fast IP/TCP/UDP checksum implementation

### PR DESCRIPTION
**Pull request by ssharks**
Instead of doing a 1-complement addition for every 16-bit word, process
32-bit words and handle the 1-complement addition at the end.
    
This work is a collaboration with Diego Pino García. The algorithm is
inspired by:
https://blogs.igalia.com/dpino/2018/06/14/fast-checksum-computation/

### CODE REVIEW

**Comment by ssharks** [comment]
@dpino: Thanks for your response, maybe you can use the implementation and unit for inspiration to fix the algorithm in your post.

**Comment by ssharks** [comment]
@rlubos: Thanks for your initial testing. I agree that misalignment could be hurting performance. Feel free to place a suggestion on how to improve there.

File: subsys/net/ip/utils.c [hunk 1]
```diff
@@ -486,32 +486,48 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
...
+ * Once the sum is computed a final step folds the sum to a 16-bit word (adding carry if any).
+ */
+uint16_t calc_chksum(uint16_t sum_in, const uint8_t *data, size_t len)
 {
-	const uint8_t *end;
-	uint16_t tmp;
-
-	end = data + len - 1;
+	uint64_t sum = (uint64_t)ntohs(sum_in);
+	uint32_t *p = (uint32_t *) data;
```
Comment (carlescufi) [inline subsys/net/ip/utils.c:499]: What about alignment? This is not gonna work on all platforms, is it?

**Review by carlescufi — CHANGES_REQUESTED** [review]
Thanks for the PR! I have a fundamental issue here with alignment, that I have posted about.

File: subsys/net/ip/utils.c (hunk 1 above)
Comment (ssharks) [reply subsys/net/ip/utils.c:499]: The sum is maintained in host endiannes, while the data is always big endian notation. By converting to big endian at the start, the whole checksum can be computed in big endian and at the end you go back to host endianness. So I think it should work both on big en little endian systems, I only tested on little endian so far.

File: subsys/net/ip/utils.c [hunk 2]
```diff
@@ -486,32 +486,48 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
...
+	uint16_t i = 0;
+	size_t pending = len;
```
Comment (rlubos) [inline subsys/net/ip/utils.c:501]: Would the following make sense?
```suggestion
	uint32_t *p;
	uint16_t i = 0;
	size_t pending = len;

	while ((((uintptr_t)data & 0x03) != 0) && (pending > 0)) {
		sum += *data;
		pending--;
		data++;
	}

	p = (uint32_t *)data;
```

That way we guarantee that `data` is aligned properly before actually using the address with `uint32_t` pointer.

File: subsys/net/ip/utils.c [hunk 3]
```diff
@@ -486,32 +486,48 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
...
+		pending -= 2;
+		sum = sum + *((uint16_t *)(p + i));
+	}
+	if (pending == 1) {
+		sum += data[len-1];
 	}
 
-	return sum;
+	/* Fold sum into 16-bit word. */
+	while (sum>>16) {
```
Comment (rlubos) [inline subsys/net/ip/utils.c:527]:
```suggestion
	while (sum >> 16) {
```

File: subsys/net/ip/utils.c [hunk 4]
```diff
@@ -486,32 +486,48 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
...
+		sum = (sum & 0xffff) + (sum>>16);
```
Comment (rlubos) [inline subsys/net/ip/utils.c:528]:
```suggestion
		sum = (sum & 0xffff) + (sum >> 16);
```

**Comment by carlescufi** [comment]
> @rlubos: Thanks for your initial testing. I agree that misalignment could be hurting performance. Feel free to place a suggestion on how to improve there.

I think this needs to be done based on platform alignment requirements, and fall back to single-byte access otherwise.

**Comment by rlubos** [comment]
CC @krish2718 @wbober

**Comment by ssharks** [comment]
> > @rlubos: Thanks for your initial testing. I agree that misalignment could be hurting performance. Feel free to place a suggestion on how to improve there.
> 
> I think this needs to be done based on platform alignment requirements, and fall back to single-byte access otherwise.

The original checksum implementation looks to me that is could have only worked on little endian architectures. This is no reason the new implementation should work in both situations. I found the CONFIG_BIG_ENDIAN configuration option. I think it makes sense run the test in both endianness configurations.

I need some rework anyway to embed @rlubos his alignment suggestions. The will mess up the endiannes anyway.

File: subsys/net/ip/utils.c (hunk 2 above)
Comment (ssharks) [reply subsys/net/ip/utils.c]: Thanks for the good hint, I took it as basis, but needed some fiddling to get the byte alignment right

File: subsys/net/ip/utils.c (hunk 3 above)
Comment (ssharks) [reply subsys/net/ip/utils.c]: Fixed

File: subsys/net/ip/utils.c (hunk 4 above)
Comment (ssharks) [reply subsys/net/ip/utils.c]: Fixed

File: subsys/net/ip/utils.c (hunk 1 above)
Comment (ssharks) [reply subsys/net/ip/utils.c]: Could you have a look if you are satisfied with the latest changes?

**Comment by ssharks** [comment]
Improved the test to certainly hit all code paths. Ready for review now.

If anyone wants to improve CRC algorithm throughput in the future, take a look at: https://create.stephan-brumme.com/crc32/

File: subsys/net/ip/utils.c [hunk 5]
```diff
@@ -486,32 +487,101 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
...
-	if (data == end) {
-		tmp = data[0] << 8;
-		sum += tmp;
-		if (sum < tmp) {
-			sum++;
-		}
+	/* Do loop unrolling for the very large data sets */
+	while (pending >= sizeof(uint32_t) * 4) {
+		uint64_t sum_a = p[i];
+		uint64_t sum_b = p[i+1];
```
Comment (rlubos) [inline subsys/net/ip/utils.c:550]:
```suggestion
		uint64_t sum_b = p[i + 1];
```

File: subsys/net/ip/utils.c [hunk 6]
```diff
@@ -486,32 +487,101 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
...
+
+		pending -= sizeof(uint32_t) * 4;
+		sum_a += p[i+2];
+		sum_b += p[i+3];
```
Comment (rlubos) [inline subsys/net/ip/utils.c:554]:
```suggestion
		sum_a += p[i + 2];
		sum_b += p[i + 3];
```

**Review by rlubos — COMMENTED** [review]
Just a few nits, looks good otherwise. Thank you for the PR!

**Comment by rlubos** [comment]
Just for reference for other reviewers, `zperf` UDP upload throughput over loopback interface (which means we test purely UDP/IP stack) running on nRF5340 platform increased from ~38 Mbps to ~47 Mbps, which is a solid improvement.

**Comment by rlubos** [comment]
> @carlescufi / @rlubos : Do you know how I can run the simulation in big endian mode? The only big endian board I found was the mips/qemu_malta. The test however did not want to run on this board, it was filtered out by the runner for some reason. Any suggestions are welcome.

It seems that this platform explicitly disables networking tests https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/mips/qemu_malta/qemu_malta_be.yaml#L14, @carlescufi do you know if we have any other options?

**Comment by jukkar** [comment]
I just tried the qemu_malta, and did this change
```
diff --git a/boards/mips/qemu_malta/qemu_malta.yaml b/boards/mips/qemu_malta/qemu_malta.yaml
index adc81837ff..b010781d3a 100644
--- a/boards/mips/qemu_malta/qemu_malta.yaml
+++ b/boards/mips/qemu_malta/qemu_malta.yaml
@@ -11,5 +11,4 @@ flash: 512
 testing:
   default: true
   ignore_tags:
-    - net
     - bluetooth
```
and I was able to run some network tests just fine. Perhaps the network support could be added back to the board.

File: subsys/net/ip/utils.c (hunk 6 above)
Comment (ssharks) [reply subsys/net/ip/utils.c]: Fixed

File: subsys/net/ip/utils.c [hunk 7]
```diff
@@ -486,32 +487,101 @@ int z_vrfy_net_addr_pton(sa_family_t family, const char *src,
...
+	 * and the offset of starting
+	 */
+	if (odd_start == CHECKSUM_BIG_ENDIAN) {
+		sum = __bswap_16(sum_in);
+	} else {
+		sum = sum_in;
+	}
 
-		data += 2;
+	/* Proess up to 3 data elements up front, so the data is aligned further down the line */
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/utils.c]: Fixed

[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/utils.c]: Fixed

**Review by rlubos — APPROVED** [review]

**Comment by rlubos** [comment]
I think the board should also specify the `netif` capability, given the test suite depends on it:
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/net/utils/testcase.yaml#L2

**Comment by ssharks** [comment]
I removed the netif requirement from the test and the net requirement from **qemu_malta_be** and was able to run the test correctly.

File: subsys/net/ip/utils.c (hunk 1 above)
Comment (ssharks) [reply subsys/net/ip/utils.c]: @carlescufi: Are you ok with the latest implementation?

**Review by carlescufi — APPROVED** [review]

**Pull request merged**